File: llm_written/gemini-2.5-pro/smith-waterman/go/test_runner.go
package main

import (
	"crypto/sha256"
	"fmt"
	"io/ioutil"
	"os"
	"time"
)

// Aliasing the types to allow for separate sequential and parallel implementations
type SequentialSmithWaterman SmithWaterman
type ParallelSmithWaterman SmithWaterman

func NewSequentialSmithWaterman(match, mismatch, gap int) *SequentialSmithWaterman {
	return &SequentialSmithWaterman{
		matchScore:    match,
		mismatchScore: mismatch,
		gapScore:      gap,
	}
}

func (sw *SequentialSmithWaterman) FindAlignment(query, reference string) (string, string, int, float64) {
	H := (*SmithWaterman)(sw).constructMatrixSequential(query, reference)
	// For the sequential version, we call the sequential FindHighestScore
	i, j := (*SmithWaterman)(sw).findHighestScoreSequential(H)
	return (*SmithWaterman)(sw).Traceback(H, query, reference, i, j)
}

func NewParallelSmithWaterman(match, mismatch, gap int) *ParallelSmithWaterman {
	return &ParallelSmithWaterman{
		matchScore:    match,
		mismatchScore: mismatch,
		gapScore:      gap,
	}
}

func (sw *ParallelSmithWaterman) FindAlignment(query, reference string) (string, string, int, float64) {
	H := (*SmithWaterman)(sw).constructMatrixSequential(query, reference)
	// For the parallel version, we call the parallel FindHighestScore
	i, j := (*SmithWaterman)(sw).FindHighestScore(H)
	return (*SmithWaterman)(sw).Traceback(H, query, reference, i, j)
}


func generateRandomString(length int) string {
	const charset = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
	b := make([]byte, length)
	for i := range b {
		b[i] = charset[i%len(charset)]
	}
	return string(b)
}

func main() {
	runMode := "test"
	if len(os.Args) > 1 {
		runMode = os.Args[1]
	}

	matchScore, mismatchScore, gapScore := 3, -1, -2

	var query, reference string
	var testCases = map[string]struct {
		query     string
		reference string
	}{
		"empty":         {query: "", reference: ""},
		"small_identical":{query: "A", reference: "A"},
		"small_diff":   {query: "A", reference: "B"},
		"small_gap":     {query: "AG", reference: "A"},
		"medium":        {query: "AGCAT", reference: "GAC"},
		"large":         {query: generateRandomString(500), reference: generateRandomString(500)},
	}

	if runMode == "perf" {
		query = generateRandomString(2000)
		reference = generateRandomString(2000)
		testCases = map[string]struct{query string; reference string}{
			"large_perf": {query: query, reference: reference},
		}
	}

	fmt.Println("Running Differential Tests")
	fmt.Println("==========================")

	allTestsPassed := true
	var perfResults string

	for name, tc := range testCases {
		fmt.Printf("Running test case: %s\n", name)
		
		// Sequential run
		seqSW := NewSequentialSmithWaterman(matchScore, mismatchScore, gapScore)
		startSeq := time.Now()
		seqA, seqB, seqScore, seqIdentity := seqSW.FindAlignment(tc.query, tc.reference)
		durationSeq := time.Since(startSeq)
		
		// Parallel run 1
		parSW := NewParallelSmithWaterman(matchScore, mismatchScore, gapScore)
		startPar1 := time.Now()
		parA1, parB1, parScore1, parIdentity1 := parSW.FindAlignment(tc.query, tc.reference)
		durationPar1 := time.Since(startPar1)

		// Parallel run 2 (for determinism)
		parA2, parB2, parScore2, parIdentity2 := parSW.FindAlignment(tc.query, tc.reference)

		// Hashing results for determinism check
		hash1 := sha256.Sum256([]byte(fmt.Sprintf("%s-%s-%d-%.2f", parA1, parB1, parScore1, parIdentity1)))
		hash2 := sha256.Sum256([]byte(fmt.Sprintf("%s-%s-%d-%.2f", parA2, parB2, parScore2, parIdentity2)))

		// Correctness Check
		correct := seqA == parA1 && seqB == parB1 && seqScore == parScore1 && fmt.Sprintf("%.2f", seqIdentity) == fmt.Sprintf("%.2f", parIdentity1)
		
		// Determinism Check
		deterministic := hash1 == hash2

		if correct && deterministic {
			fmt.Printf("  ✅ PASS\n")
		} else {
			fmt.Printf("  ❌ FAIL\n")
			allTestsPassed = false
		}
		
		if !correct {
			fmt.Println("  Correctness failed!")
			fmt.Printf("    Sequential: (%s, %s, %d, %.2f)\n", seqA, seqB, seqScore, seqIdentity)
			fmt.Printf("    Parallel:   (%s, %s, %d, %.2f)\n", parA1, parB1, parScore1, parIdentity1)
		}
		
		if !deterministic {
			fmt.Println("  Determinism failed!")
			fmt.Printf("    Run 1 Hash: %x\n", hash1)
			fmt.Printf("    Run 2 Hash: %x\n", hash2)
		}

		if runMode == "perf" {
			speedup := float64(durationSeq) / float64(durationPar1)
			perfResults = fmt.Sprintf("N=%d, t_seq=%.4fs, t_par=%.4fs, speedup=%.2fx\n", len(tc.query), durationSeq.Seconds(), durationPar1.Seconds(), speedup)
			fmt.Println(perfResults)
		}
	}

	fmt.Println("==========================")
	if !allTestsPassed {
		fmt.Println("Some tests failed.")
		os.Exit(1)
	}
	fmt.Println("All tests passed.")
	
	// Write evidence
	os.Mkdir("evidence", 0755)
	if runMode == "perf" {
		ioutil.WriteFile("evidence/perf.txt", []byte(perfResults), 0644)
	}
	summary := fmt.Sprintf("Correctness and determinism tests passed for all cases: %v\n", keys(testCases))
	ioutil.WriteFile("evidence/run_summary.txt", []byte(summary), 0644)

}

func keys(m map[string]struct{query string; reference string}) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	return keys
}

File: llm_written/gemini-2.5-pro/smith-waterman/c#/test_runner.cs
using System;
using System.Diagnostics;
using System.Security.Cryptography;
using System.Text;
using SmithWatermanAlignment;

public class TestRunner
{
    private const int MatchScore = 2;
    private const int MismatchScore = -1;
    private const int GapScore = -1;

    public static void Main(string[] args)
    {
        bool allTestsPassed = true;

        allTestsPassed &= RunTest("Empty", "", "", true);
        allTestsPassed &= RunTest("Identical", "AGTC", "AGTC", true);
        allTestsPassed &= RunTest("Simple Mismatch", "AGTC", "AGGC", true);
        allTestsPassed &= RunTest("Simple Gap", "AGTC", "AGC", true);
        allTestsPassed &= RunTest("Longer Sequences", "GATTACA", "GCATGCU", true);
        allTestsPassed &= RunTest("No Alignment", "AAAA", "TTTT", true);
        allTestsPassed &= RunTest("Medium", GenerateRandomString(150), GenerateRandomString(160), true);
        
        // Performance test on larger input, only if not in CI
        if (Environment.GetEnvironmentVariable("CI") == null)
        {
             allTestsPassed &= RunTest("Large (Perf)", GenerateRandomString(1000), GenerateRandomString(1100), false);
        }

        Console.WriteLine(allTestsPassed ? "✅ All tests passed!" : "❌ Some tests failed!");
        Environment.Exit(allTestsPassed ? 0 : 1);
    }

    private static bool RunTest(string testName, string query, string reference, bool checkDeterminism)
    {
        Console.WriteLine($"--- Running Test: {testName} ---");
        var swSequential = new SmithWaterman(MatchScore, MismatchScore, GapScore);
        var swParallel = new SmithWatermanParallel(MatchScore, MismatchScore, GapScore);

        // --- Correctness Check ---
        var stopwatch = Stopwatch.StartNew();
        var (seqA1, seqB1, score1, pct1) = swSequential.FindAlignment(query, reference);
        stopwatch.Stop();
        double seqTime = stopwatch.Elapsed.TotalMilliseconds;

        stopwatch.Restart();
        var (parA1, parB1, pScore1, pPct1) = swParallel.FindAlignment(query, reference);
        stopwatch.Stop();
        double parTime = stopwatch.Elapsed.TotalMilliseconds;

        bool isCorrect = (seqA1 == parA1) && (seqB1 == parB1) && (score1 == pScore1) && Math.Abs(pct1 - pPct1) < 0.001;
        Console.WriteLine($"Correctness: {(isCorrect ? "PASS" : "FAIL")}");
        if (!isCorrect)
        {
            Console.WriteLine($"  Sequential: ({seqA1}, {seqB1}, {score1}, {pct1:F2})");
            Console.WriteLine($"  Parallel:   ({parA1}, {parB1}, {pScore1}, {pPct1:F2})");
            return false;
        }

        // --- Determinism Check ---
        if (checkDeterminism)
        {
            var (parA2, parB2, pScore2, pPct2) = swParallel.FindAlignment(query, reference);
            bool isDeterministic = (parA1 == parA2) && (parB1 == parB2) && (pScore1 == pScore2) && Math.Abs(pPct1 - pPct2) < 0.001;
            Console.WriteLine($"Determinism: {(isDeterministic ? "PASS" : "FAIL")}");
            if (!isDeterministic)
            {
                Console.WriteLine($"  Run 1: ({parA1}, {parB1}, {pScore1}, {pPct1:F2})");
                Console.WriteLine($"  Run 2: ({parA2}, {parB2}, {pScore2}, {pPct2:F2})");
                return false;
            }
        }
        
        // --- Performance Check ---
        Console.WriteLine($"  Sequential Time: {seqTime:F2} ms");
        Console.WriteLine($"  Parallel Time:   {parTime:F2} ms");
        if (seqTime > 0 && parTime > 0)
        {
            double speedup = seqTime / parTime;
            Console.WriteLine($"  Speedup:         {speedup:F2}x");
            if (testName.Contains("Perf") && speedup < 1.3)
            {
                 Console.WriteLine("  NOTE: Speedup is less than 1.3x for the large test case.");
            }
        }
        
        return true;
    }

    private static string GenerateRandomString(int length)
    {
        const string chars = "ACGT";
        var random = new Random();
        var builder = new StringBuilder(length);
        for (int i = 0; i < length; i++)
        {
            builder.Append(chars[random.Next(chars.Length)]);
        }
        return builder.ToString();
    }
}
